Version detection

PanicSheep opened the pull request:
@AnnikaLau Hi Annika, I saw that we're working on the same feature, related to #367. So I branched off of your branch `release_detection`, added my ideas and opened this PR draft so we can collaborate.

PanicSheep commented:
launch jenkins spack info cosmo

PanicSheep commented:
```
MakefilePackage:   cosmo

Description:
   ​COSMO: Numerical Weather Prediction Model. Needs access to private
   ​GitHub.

Homepage: http://www.cosmo-model.org

Maintainers: @elsagermann

Externally Detectable: 
   ​False

Tags: 
   ​None

Preferred version:  
   ​mch                            [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch

Safe versions:  
   ​master                         [git] git@github.com:COSMO-ORG/cosmo.git on branch master
   ​org-master                     [git] git@github.com:COSMO-ORG/cosmo.git on branch master
   ​mch                            [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch
   ​dev-build                      [git] git@github.com:COSMO-ORG/cosmo.git on branch master
   ​c2sm-master                    [git] git@github.com:C2SM-RCM/cosmo.git on branch master
   ​c2sm_5.09a                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.09a
   ​c2sm_5.09                      [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.09
   ​c2sm_5.08                      [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.08
   ​c2sm_5.07c                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.07c
   ​c2sm_5.07b                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.07b
   ​c2sm_5.07a                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.07a
   ​c2sm_5.07                      [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.07
   ​c2sm_5.06_1                    [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.06_1
   ​c2sm_5.06b                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.06b
   ​c2sm_5.06a                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.06a
   ​c2sm_5.05b                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.05b
   ​c2sm_5.05a_1                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.05a_1
   ​c2sm_5.05a                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.05a
   ​c2sm_5.05                      [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.05
   ​c2sm_5.04h                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04h
   ​c2sm_5.04f                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04f
   ​c2sm_5.04e1                    [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04e1
   ​c2sm_5.04d3                    [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04d3
   ​c2sm_5.04d2                    [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04d2
   ​c2sm_5.04c1                    [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04c1
   ​c2sm_5.04a                     [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04a
   ​c2sm_5.04                      [git] git@github.com:C2SM-RCM/cosmo.git at tag 5.04
   ​c2sm_1.04.16                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.16
   ​c2sm_1.04.15                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.15
   ​c2sm_1.04.14                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.14
   ​c2sm_1.04.12                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.12
   ​c2sm_1.04.11                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.11
   ​c2sm_1.04.10                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.10
   ​c2sm_1.04.09                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.09
   ​c2sm_1.04.08                   [git] git@github.com:C2SM-RCM/cosmo.git at tag 1.04.08
   ​c2sm-features                  [git] git@github.com:C2SM-RCM/cosmo.git on branch c2sm-features
   ​apn_5.14.mch1.0                [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.14.mch1.0
   ​apn_5.09a.mch1.2.p1            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.09a.mch1.2.p1
   ​apn_5.09a.mch1.1.p1            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.09a.mch1.1.p1
   ​apn_5.09a.mch1.0.p1            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.09a.mch1.0.p1
   ​apn_5.09a.mch1.0.base          [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.09a.mch1.0.base
   ​apn_5.08.mch1.1                [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch1.1
   ​apn_5.08.mch.1.0.p6            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch.1.0.p6
   ​apn_5.08.mch.1.0.p5            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch.1.0.p5
   ​apn_5.08.mch.1.0.p4            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch.1.0.p4
   ​apn_5.08.mch.1.0.p3            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch1.0.p3
   ​apn_5.08.mch1.0.p2             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch1.0.p2
   ​apn_5.08.mch1.0.p1             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch1.0.p1
   ​apn_5.08.mch1.0.base           [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.08.mch1.0.base
   ​apn_5.07.mch1.0.p12            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p12
   ​apn_5.07.mch1.0.p11            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p11
   ​apn_5.07.mch1.0.p10            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p10
   ​apn_5.07.mch1.0.p9             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p9
   ​apn_5.07.mch1.0.p8             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p8
   ​apn_5.07.mch1.0.p7             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p7
   ​apn_5.07.mch1.0.p6             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p6
   ​apn_5.07.mch1.0.p5             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p5
   ​apn_5.07.mch1.0.p4             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p4
   ​apn_5.07.mch1.0.p3             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p3
   ​apn_5.07.mch1.0.p2             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p2
   ​apn_5.07.mch1.0.p1             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p1
   ​apn_5.07.mch1.0.p0             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.p0
   ​apn_5.07.mch1.0.base           [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07.mch1.0.base
   ​apn_5.07a.mch1.0.p1            [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07a.mch1.0.p1
   ​apn_5.07a.mch1.0.base          [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.07a.mch1.0.base
   ​apn_5.06b.mch1.5               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.5
   ​apn_5.06b.mch1.4               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.4
   ​apn_5.06b.mch1.3               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.3
   ​apn_5.06b.mch1.2               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.2
   ​apn_5.06b.mch1.1.pgifix.tsa    [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.1.pgifix.tsa
   ​apn_5.06b.mch1.1.pgifix        [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.1.pgifix
   ​apn_5.06b.mch1.1               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.1
   ​apn_5.06b.mch1.0               [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06b.mch1.0
   ​apn_5.06a.mch1.rc1             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06a.mch1.rc1
   ​apn_5.06a.mch-gt2.rc1          [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.06a.mch-gt2.rc1
   ​apn_5.05b.mch1.rc2             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.05b.mch1.rc2
   ​apn_5.05b.mch1.rc1             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag 5.05b.mch1.rc1
   ​apn_mchrc_06122019             [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag mchrc_06122019
   ​apn-mch                        [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch
   ​apn_archive_mch_20201103       [git] git@github.com:MeteoSwiss-APN/cosmo.git at tag archive_mch_20201103

Deprecated versions:  
   ​None

Variants:
   ​Name [Default]               When    Allowed values    Description
   ​=========================    ====    ==============    ===========================

   ​claw [off]                   --      on, off           Build with claw-compiler
   ​cosmo_target [gpu]           --      gpu, cpu          Build with target gpu or
                                                          ​cpu
   ​cppdycore [on]               --      on, off           Build with the C++ DyCore
   ​cuda_arch [none]             --      70, 60, 37        Build with cuda_arch
   ​debug [off]                  --      on, off           Build debug mode
   ​dycoretest [on]              --      on, off           Build C++ dycore with
                                                          ​testing
   ​eccodes [on]                 --      on, off           Build with eccodes instead
                                                          ​of grib-api
   ​gt1 [off]                    --      on, off           Build dycore with gridtools
                                                          ​1.1.3
   ​oasis [off]                  --      on, off           Build with the unified
                                                          ​oasis interface
   ​parallel [on]                --      on, off           Build parallel COSMO
   ​pollen [off]                 --      on, off           Build with pollen enabled
   ​production [off]             --      on, off           Force all variants to be
                                                          ​the ones used in production
   ​real_type [double]           --      double, float     Build with double or single
                                                          ​precision enabled
   ​serialize [off]              --      on, off           Build with serialization
                                                          ​enabled
   ​set_version [off]            --      on, off           Pass cosmo tag version to
                                                          ​Makefile
   ​slave [tsa]                  --                        Build on slave tsa, daint
                                                          ​or kesch
   ​slurm_account [g110]         --                        Slurm option to specify
                                                          ​account for testing
   ​slurm_bin [srun]             --                        Slurm binary on CSCS
                                                          ​machines
   ​slurm_constraint [gpu]       --                        Slurm constraints for nodes
                                                          ​requested
   ​slurm_nodes [1]              --                        Pattern to specify number
                                                          ​of nodes for testing
   ​slurm_opt_account [-A]       --                        Slurm option to specify
                                                          ​account for testing
   ​slurm_opt_constraint [-C]    --                        Slurm option to specify
                                                          ​constraints for nodes
                                                          ​requested
   ​slurm_opt_nodes [-n]         --                        Slurm option to specify
                                                          ​number of nodes for testing
   ​slurm_opt_partition [-p]     --                        Slurm option to specify
                                                          ​partition for testing
   ​slurm_partition [normal]     --                        Slurm partition for testing
   ​verbose [off]                --      on, off           Build cosmo with verbose
                                                          ​enabled
   ​zlib_ng [off]                --      on, off           Run with faster zlib-
                                                          ​implemention for
                                                          ​compression of NetCDF
                                                          ​output

Installation Phases:
   ​edit    build    install

Build Dependencies:
   ​boost
   ​claw
   ​cmake
   ​cosmo-dycore
   ​cosmo-eccodes-definitions
   ​cosmo-grib-api-definitions
   ​cuda
   ​jasper
   ​libgrib1
   ​mpi
   ​mpicuda
   ​netcdf-c
   ​netcdf-fortran
   ​oasis
   ​omni-xmod-pool
   ​perl
   ​serialbox

Link Dependencies:
   ​cosmo-eccodes-definitions
   ​cuda
   ​mpi
   ​mpicuda
   ​netcdf-c
   ​netcdf-fortran
   ​oasis
   ​zlib_ng

Run Dependencies:
   ​cosmo-eccodes-definitions
   ​cosmo-grib-api-definitions
   ​cuda
   ​mpi
   ​mpicuda
   ​oasis
   ​slurm
   ​zlib_ng

Virtual Packages: 
   ​None
```

PanicSheep commented:
@AnnikaLau Why did you add the prefixes to the tags? Was there a clash?

PanicSheep commented:
```
Warning: no access to git@github.com:MeteoSwiss-APN/cosmo.git => not fetching versions


Warning: no access to git@github.com:C2SM-RCM/cosmo.git => not fetching versions

MakefilePackage:   cosmo

Description:
    COSMO: Numerical Weather Prediction Model. Needs access to private
    GitHub.

Homepage: http://www.cosmo-model.org

Maintainers: @elsagermann

Externally Detectable: 
    False

Tags: 
    None

Preferred version:  
    mch              [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch

Safe versions:  
    master           [git] git@github.com:COSMO-ORG/cosmo.git on branch master
    org-master       [git] git@github.com:COSMO-ORG/cosmo.git on branch master
    mch              [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch
    dev-build        [git] git@github.com:COSMO-ORG/cosmo.git on branch master
    c2sm-master      [git] git@github.com:C2SM-RCM/cosmo.git on branch master
    c2sm-features    [git] git@github.com:C2SM-RCM/cosmo.git on branch c2sm-features
    apn-mch          [git] git@github.com:MeteoSwiss-APN/cosmo.git on branch mch

Deprecated versions:  
    None

Variants:
    Name [Default]               When    Allowed values    Description
    =========================    ====    ==============    ==================================================================

    claw [off]                   --      on, off           Build with claw-compiler
    cosmo_target [gpu]           --      gpu, cpu          Build with target gpu or cpu
    cppdycore [on]               --      on, off           Build with the C++ DyCore
    cuda_arch [none]             --      70, 60, 37        Build with cuda_arch
    debug [off]                  --      on, off           Build debug mode
    dycoretest [on]              --      on, off           Build C++ dycore with testing
    eccodes [on]                 --      on, off           Build with eccodes instead of grib-api
    gt1 [off]                    --      on, off           Build dycore with gridtools 1.1.3
    oasis [off]                  --      on, off           Build with the unified oasis interface
    parallel [on]                --      on, off           Build parallel COSMO
    pollen [off]                 --      on, off           Build with pollen enabled
    production [off]             --      on, off           Force all variants to be the ones used in production
    real_type [double]           --      double, float     Build with double or single precision enabled
    serialize [off]              --      on, off           Build with serialization enabled
    set_version [off]            --      on, off           Pass cosmo tag version to Makefile
    slave [tsa]                  --                        Build on slave tsa, daint or kesch
    slurm_account [g110]         --                        Slurm option to specify account for testing
    slurm_bin [srun]             --                        Slurm binary on CSCS machines
    slurm_constraint [gpu]       --                        Slurm constraints for nodes requested
    slurm_nodes [1]              --                        Pattern to specify number of nodes for testing
    slurm_opt_account [-A]       --                        Slurm option to specify account for testing
    slurm_opt_constraint [-C]    --                        Slurm option to specify constraints for nodes requested
    slurm_opt_nodes [-n]         --                        Slurm option to specify number of nodes for testing
    slurm_opt_partition [-p]     --                        Slurm option to specify partition for testing
    slurm_partition [normal]     --                        Slurm partition for testing
    verbose [off]                --      on, off           Build cosmo with verbose enabled
    zlib_ng [off]                --      on, off           Run with faster zlib-implemention for compression of NetCDF output

Installation Phases:
    edit    build    install

Build Dependencies:
    boost  cmake         cosmo-eccodes-definitions   cuda    libgrib1  mpicuda   netcdf-fortran  omni-xmod-pool  serialbox
    claw   cosmo-dycore  cosmo-grib-api-definitions  jasper  mpi       netcdf-c  oasis           perl

Link Dependencies:
    cosmo-eccodes-definitions  cuda  mpi  mpicuda  netcdf-c  netcdf-fortran  oasis  zlib_ng

Run Dependencies:
    cosmo-eccodes-definitions  cosmo-grib-api-definitions  cuda  mpi  mpicuda  oasis  slurm  zlib_ng

Virtual Packages: 
    None

```

AnnikaLau commented:
> @AnnikaLau Why did you add the prefixes to the tags? Was there a clash?
@PanicSheep : Hi Dominic, I added the prefixes to the tags because otherwise you wouldn't see that the same tags are available in different repos. The version function overwrites metadata for the same tag.

PanicSheep commented:
Wow, I just realized that without this PR, introducing a new tag in one repo might disrupt someone working with an other repo, just because the tags are named the same. That's a rather fragile design.
So I think we should proceed with this new prefix although it implies that users need to adapt to the new naming convention once we merge this PR.

TODO:
- [ ] Inform users that they need to adapt to the new naming convention

AnnikaLau commented:
@PanicSheep: Are you still planning to implement any features here or should we proceed with a review request? It would be fine from my side.

PanicSheep commented:
@AnnikaLau Can you move the file `tools/spack-scripting/scripting/cmd/version_detection.py` to `tools/version_detection.py`? All the other files in `tools/spack-scripting/scripting/cmd/` are related to command line extensions of spack, yet version detection is not. Feel free to push that change to this branch. Once done, could you trigger jenkins? We changed the behavior of cosmo, icontools and int2lm, so I think we should test them and ignore the white space changes in other packages. This should do it: `launch jenkins cosmo icontools int2lm`. Do you agree? I have nothing else to add to this PR. I think it's fine if you and I review. 4 eyes should be enough.
I added this PR as an [agenda](https://github.com/C2SM/spack-c2sm/wiki/Minutes-and-agenda-of-spack-team-meetings) point so we can discuss if we should inform users.

AnnikaLau commented:
launch jenkins cosmo icontools int2lm

AnnikaLau commented:
launch jenkins cosmo icontools int2lm

AnnikaLau commented:
@PanicSheep : I moved the version_detection.py as you suggested. I also think the PR is fine otherwise. Thanks for reviewing!

PanicSheep commented:
There were conflicts with master. I resolved them.

PanicSheep commented:
launch jenkins cosmo icontools int2lm

AnnikaLau reviewed (approved):
For me everything looks ok now.

AnnikaLau commented:
@PanicSheep : do you want to do one last review?
